### PR TITLE
fix(fakeAsync): have fakeAsync use Proxy zone.

### LINKED ISSUE #10503: When `inject` is before `fakeAsync` (and probably async), tests do not use the expected zone
When calling `TestBed.createComponent` or `TestCombponentBuilder.createAsync`, [the function uses `injector.get` to retrieve the `NgZone`](https://github.com/angular/angular/blob/master/modules/@angular/core/testing/test_bed.ts#L355). Any tasks (microTasks or macroTasks) that occur as part of the `createComponent` are added to this zone (and not to `Zone.current`). 

The `NgZone` is instantiated when `inject` is called, using whatever zone is `Zone.current` when the [constructor](https://github.com/angular/angular/blob/master/modules/@angular/core/src/zone/ng_zone_impl.ts#L47) is run.

This means that when `inject` is called _outside_ of a `fakeAsync` or `async`, then `Zone.current` will be `"angular"` (direct child of `"<root>"`).

As a result of this, any tasks spawned from calling `createComponent` will be added to the `"angular"` zone _instead_ of the testing zone when the test uses either `async` or `fakeAsync`.

Example:
```ts
beforEach(inject((tcb: TestComponentBuilder) => {
  builder = tcb;
}));

it('should test something with fakeAsync', fakeAsync(() => {
  let appFixture: ComponentFixture<any>;
  builder.createAsync(SomeComponent).then(fixture => {
    appFixture = fixture;
  });

  flushMicrotasks();

  expect(appFixture).toBeTruthy(); // FAIL, appFixture is undefined
}));
```

Here, the call to `flushMicrotasks` will *not* cause the `then`'ed function to execute because the task for the `createAsync` was added to the `"angular"` zone and *not* the `fakeAsyncTestZone`.

As a workaround, you can write:
```ts
  beforeEach(fakeAsync(inject([TestComponentBuilder], (tcb: TestComponentBuilder) => {
    builder = tcb;
  })));
```
When the function returned from `fakeAsync` is called, the current zone is forked into the `fakeAsyncTestZone`, so that `Zone.current` is the expected zone when `inject` runs and instantiates the `NgZone`, which is then consumed by `createAsync`, causing the microtasks to be added to the expected zone.

cc @IgorMinar @mhevery @juliemr

### PR DESCRIPTION
**Pull request by mhevery**
Fixes #10503

The PR fails until https://github.com/angular/zone.js/pull/396 is merged and pushed to NPM

### CODE REVIEW

**Comment by vikerman** [comment]
Can we add a test that exercises the exact problem in the original bug?

**Comment by mhevery** [comment]
@vikerman Added test per your request. (https://github.com/angular/angular/pull/10797/files#diff-f24a03f8500f7f8fe3679340dc86b890R758)

File: karma-js.conf.js [hunk 1]
```diff
@@ -6,7 +6,7 @@ var internalAngularReporter = require('./tools/karma/reporter.js');
 module.exports = function(config) {
   config.set({
 
-    frameworks: ['jasmine'],
+    frameworks: ['jasmine', 'zonejs'],
```
Comment (juliemr) [inline karma-js.conf.js]: Please push `karma-zonejs` master to github :)

File: karma-js.conf.js [hunk 2]
```diff
@@ -17,12 +17,6 @@ module.exports = function(config) {
       // include Angular v1 for upgrade module testing
```
Comment (juliemr) [inline karma-js.conf.js]: I have a vague memory of zonejs failing if it's loaded _before_ es6-shim. Can we confirm that this isn't a problem with `zonejs` as a framework plugin? Maybe I'm misremembering?

File: modules/@angular/platform-browser-dynamic/test/testing_public_browser_spec.ts [hunk 3]
```diff
@@ -100,8 +100,12 @@ export function main() {
         var itPromise = patchJasmineIt();
 
         it('should fail with an error from a promise', async(() => {
-             TestBed.configureTestingModule({declarations: [BadTemplateUrl]});
-             TestBed.compileComponents();
+             try {
+               TestBed.configureTestingModule({declarations: [BadTemplateUrl]});
+               TestBed.compileComponents();
+             } catch (e) {
```
Comment (juliemr) [inline modules/@angular/platform-browser-dynamic/test/testing_public_browser_spec.ts]: Probably should be removed.

File: test-main.js [hunk 4]
```diff
@@ -68,40 +68,51 @@ System.config({
   }
 });
```
Comment (juliemr) [inline test-main.js]: Probably need to change test-cjs-main as well.

File: test-main.js [hunk 5]
```diff
@@ -68,40 +68,51 @@ System.config({
...
-})
-.then(function() {
-  __karma__.start();
-}, function(error) {
-  __karma__.error(error.stack || error);
+var ProxyZoneSpec = Zone['ProxyZoneSpec'];
+// IMPORTANT: In order for the fakeAsync tests to work all of the test code needs to
+// run in the ProxyZone regardless of if the test uses fakeAsync. The easiest way to
+// achieve this is to run everything in the the ProxyZone.
+var warnOnAsync = {
```
Comment (juliemr) [inline test-main.js]: Debugging? Remove?

**Comment by juliemr** [comment]
I think we still need to deal with `async` helper as well, or we'll continue to get the problem described in https://github.com/angular/angular/issues/10503#issuecomment-237672010

File: karma-js.conf.js (hunk 2 above)
Comment (mhevery) [inline karma-js.conf.js]: Your memory serves you well. See here: https://github.com/angular/zone.js/pull/396/commits/887acf5882c6d2d31e2acb25a8777e94cabc86dc

File: karma-js.conf.js (hunk 1 above)
Comment (mhevery) [inline karma-js.conf.js]: done

File: modules/@angular/platform-browser-dynamic/test/testing_public_browser_spec.ts (hunk 3 above)
Comment (mhevery) [inline modules/@angular/platform-browser-dynamic/test/testing_public_browser_spec.ts]: done

File: test-main.js (hunk 5 above)
Comment (mhevery) [inline test-main.js]: removed

**Comment by mhevery** [comment]
@juliemr I have added tests as requested. 

This PR will not be green until https://github.com/angular/zone.js/pull/402 is merged and pushed to NPM

File: modules/@angular/core/test/fake_async_spec.ts [hunk 6]
```diff
@@ -294,11 +295,26 @@ export function main() {
...
            expect(zoneInTest1).toBe(zoneInBeforeEach);
          }));
+    });
+  });
 
-      it('should use a different zone between tests', fakeAsync(() => {
-           expect(Zone.current).toBe(zoneInBeforeEach);
-           expect(Zone.current).not.toBe(zoneInTest1);
-         }));
+  ddescribe('ProxyZone', () => {
```
Comment (IgorMinar) [inline modules/@angular/core/test/fake_async_spec.ts]: ddescribe

File: modules/@angular/core/test/fake_async_spec.ts [hunk 7]
```diff
@@ -294,11 +295,26 @@ export function main() {
...
+    beforeEach(() => { ProxyZoneSpec.assertPresent(); });
+
+    afterEach(() => { ProxyZoneSpec.assertPresent(); });
+
+    it('should allow fakeAsync zone to retroactivly set a zoneSpec outside of fakeAsync', () => {
```
Comment (IgorMinar) [inline modules/@angular/core/test/fake_async_spec.ts]: retroactively

File: modules/@angular/core/test/fake_async_spec.ts [hunk 8]
```diff
@@ -294,11 +295,26 @@ export function main() {
...
+      ProxyZoneSpec.assertPresent();
+      var state: string = 'not run';
+      const testZone = Zone.current.fork({name: 'test-zone'});
+      (fakeAsync(() => {
+        testZone.run(() => {
+          Promise.resolve('works').then((v) => state = v);
+          expect(state).toEqual('not run');
+          flushMicrotasks();
+          expect(state).toEqual('works');
```
Comment (IgorMinar) [inline modules/@angular/core/test/fake_async_spec.ts]: can you move this to the bottom of this spec outside of fakeAsync to verify that it actually executed?

File: tools/cjs-jasmine/index.ts [hunk 9]
```diff
@@ -65,11 +66,17 @@ jrunner.configureDefaultReporter({showColors: process.argv.indexOf('--no-color')
...
-require('zone.js/dist/jasmine-patch.js');
-distAllRequire('@angular/platform-server/src/parse5_adapter.js').Parse5DomAdapter.makeCurrent();
-specFiles.forEach((file: string) => {
-  var r = distAllRequire(file);
-  if (r.main) r.main();
+var ProxyZoneSpec: {new (): ZoneSpec} = (Zone as any)['ProxyZoneSpec'];
+// IMPORTANT: In order for the fakeAsync tests to work all of the test code needs to
+// run in the ProxyZone regardless of if the test uses fakeAsync. The easiest way to
+// achieve this is to run everything in the the ProxyZone.
+Zone.current.fork(new ProxyZoneSpec()).run(function() {
```
Comment (IgorMinar) [inline tools/cjs-jasmine/index.ts]: @jasonchoimtt this change will impact your fork - FYI.

Comment (IgorMinar) [inline tools/cjs-jasmine/index.ts]: @mhevery does the zone forking in Karma happen in the karma zone framework? where is the code for that?

File: modules/@angular/core/test/fake_async_spec.ts (hunk 6 above)
Comment (mhevery) [inline modules/@angular/core/test/fake_async_spec.ts]: fixed

File: modules/@angular/core/test/fake_async_spec.ts (hunk 7 above)
Comment (mhevery) [inline modules/@angular/core/test/fake_async_spec.ts]: fixed

File: modules/@angular/core/test/fake_async_spec.ts (hunk 8 above)
Comment (mhevery) [inline modules/@angular/core/test/fake_async_spec.ts]: done

File: tools/cjs-jasmine/index.ts (hunk 9 above)
Comment (mhevery) [inline tools/cjs-jasmine/index.ts]: This needs to be reverted.

File: modules/@angular/core/testing/async.ts [hunk 10]
```diff
@@ -50,13 +50,46 @@ export function async(fn: Function): (done: any) => any {
...
+  };
+  if (ProxyZoneSpec === undefined) {
+    throw new Error(
+        'ProxyZoneSpec is needed for the async() test helper but could not be found. ' +
+        'Please make sure that your environment includes zone.js/dist/proxy-zone.js');
+  }
+  const proxyZoneSpec = ProxyZoneSpec.get();
+  ProxyZoneSpec.assertPresent();
+  // We need to create the AsyncTestZoneSpec outside the ProxyZone.
+  // If we do in in ProxyZone then we will get to infinite recursion.
```
Comment (IgorMinar) [inline modules/@angular/core/testing/async.ts]: in -> it

File: modules/@angular/core/testing/fake_async.ts [hunk 11]
```diff
@@ -45,26 +50,31 @@ let _inFakeAsyncCall = false;
  * @experimental
  */
 export function fakeAsync(fn: Function): (...args: any[]) => any {
-  return function(...args: any[] /** TODO #9100 */) {
+  return function(...args: any[]) {
+    const proxyZoneSpec = ProxyZoneSpec.assertPresent();
     if (_inFakeAsyncCall) {
       throw new BaseException('fakeAsync() calls can not be nested');
```
Comment (IgorMinar) [inline modules/@angular/core/testing/fake_async.ts]: do you still need this check? seems redundant given the check right below.

File: modules/@angular/core/testing/fake_async.ts [hunk 12]
```diff
@@ -45,26 +50,31 @@ let _inFakeAsyncCall = false;
...
+      const lastProxyZoneSpec = proxyZoneSpec.getDelegate();
+      proxyZoneSpec.setDelegate(_fakeAsyncTestZoneSpec);
+      try {
+        res = fn(...args);
         flushMicrotasks();
-        return res;
-      });
+      } finally {
+        proxyZoneSpec.resetDelegate();
+        proxyZoneSpec.setDelegate(lastProxyZoneSpec);
```
Comment (IgorMinar) [inline modules/@angular/core/testing/fake_async.ts]: it's looks suspicious that you need to both reset and set the delegate. why? can we do betteR?

**Comment by IgorMinar** [comment]
ci still unhappey

File: modules/@angular/core/testing/async.ts (hunk 10 above)
Comment (mhevery) [inline modules/@angular/core/testing/async.ts]: fixed

File: modules/@angular/core/testing/fake_async.ts (hunk 11 above)
Comment (mhevery) [inline modules/@angular/core/testing/fake_async.ts]: it is also used to retrieve the reference.

File: modules/@angular/core/testing/fake_async.ts (hunk 12 above)
Comment (mhevery) [inline modules/@angular/core/testing/fake_async.ts]: you are right reset is not necessary. Removed.

**Comment by juliemr** [comment]
For the record (we've discussed offline), I think this needs additional fixes like https://github.com/juliemr/angular/commit/6b39322bcd09e059bf5fef40adea8c33df9223c1 to solve timeout issues.

**Comment by StefH** [comment]
Is this also related to this issue ?
https://github.com/angular/angular/issues/11317

**Pull request merged**